examples/vae_text/vae_train.py

```diff
@@ -0,0 +1,463 @@
+
+    save_path = os.path.join(save_dir, suffix)
+
+    # KL term annealing rate
+    anneal_r = 1.0 / (config.kl_anneal_hparams["warm_up"] *
+        (train_data.__len__() / config.batch_size))
+
+    model = Vae(train_data)
+    model.to(device)
+
+    optimizer = tx.core.get_optimizer(params=model.parameters(),
+                                    hparams=config.opt_hparams)
+
+    def _run_epoch(epoch, mode, display=10):
+        if mode == 'train':
+            iterator.switch_to_train_data()
+        elif mode == 'valid':
+            iterator.switch_to_val_data()
+        elif mode == 'test':
+            iterator.switch_to_test_data()
+        if mode == 'train':
+            model.train()
+        else:
+            model.eval()
+        step = 0
+        start_time = time.time()
+        num_words = num_sents = 0
+        nll_ = 0.
+        kl_loss_ = rc_loss_ = 0.
+
+        for batch in iterator:
+            if mode == 'train':
+                opt_vars["kl_weight"] = min(
+                    1.0, opt_vars["kl_weight"] + anneal_r)
+
+                kl_weight_ = opt_vars["kl_weight"]
+            else:
+                kl_weight_ = 1.0
+
+            mode = mode
+
+            if mode == "train":
+                adjust_learning_rate(optimizer, opt_vars["learning_rate"])
+
+            fetches_ = model(batch, kl_weight_)
+            if mode == "train":
+                fetches_["nll"].backward()
+                optimizer.step()
+                optimizer.zero_grad()
+
+            batch_size_ = len(fetches_["lengths"])
+            num_sents += batch_size_
+            num_words += sum(fetches_["lengths"])
+            nll_ += fetches_["nll"].item() * batch_size_
+            kl_loss_ += fetches_["kl_loss"] * batch_size_
+            rc_loss_ += fetches_["rc_loss"] * batch_size_
+
+            if step % display == 0 and mode == 'train':
+                nll = nll_ / float(num_sents)
+                klw = opt_vars["kl_weight"]
+                KL = kl_loss_ / float(num_sents)
+                rc = rc_loss_ / float(num_sents)
+                log_ppl = nll_ / float(num_words)
+                ppl = np.exp(log_ppl)
+                time_cost = time.time() - start_time
+
+                print(f"{mode}: epoch {epoch}, step {step}, nll {nll:.4f}, "
+                      f"klw {klw:.4f}, KL {KL:.4f}, rc {rc:.4f}, "
+                      f"log_ppl {log_ppl:.4f}, ppl {ppl:.4f}, "
+                      f"time_cost {time_cost:.1f}")
+                sys.stdout.flush()
+
+            step += 1
+
+        nll = nll_ / float(num_sents)
+        KL = kl_loss_ / float(num_sents)
+        rc = rc_loss_ / float(num_sents)
+        log_ppl = nll_ / float(num_words)
+        ppl = np.exp(log_ppl)
+        print(f"\n{mode}: epoch {epoch}, nll {nll:.4f}, KL {KL:.4f}, "
+              f"rc {rc:.4f}, log_ppl {log_ppl:.4f}, ppl {ppl:.4f}")
+        return nll, ppl
+
+    @torch.no_grad()
+    def _generate(fname=None):
```

Comment:
Call this `filename` or `path` instead of ambiguous `fname`.

Reply:
Corrected.